core/arch/arm/kernel/thread_a32.S

```diff
@@ -63,6 +63,13 @@ UNWIND(	.cantunwind)
 	bx	lr
 END_FUNC thread_set_fiq_sp
 
+FUNC thread_a32_get_usr_sp , :
+	cps	#CPSR_MODE_SYS
```

Comment:
Is it save to switch to system mode anywhere? My primary concern is with interrupts enabled and/or in paged code.

Reply:
The function name is misleading. I wanted the assembly function as simple as possible to read out **usr_sp**. Caller should call this only from safe contexts of course.

Reply:
Yeah, so that leads me to the question, what is a safe context for this function? :-)
It might be safe both in paged code and with interrupts unmasked, it takes a bit of digging to be sure.

Reply:
I assume that these updates means that it's not safe to take non-secure interrupts while executing this function.
How about page faults?

Reply:
I made this function similar to the other `thread_set_xxx()` here. I assume caller knows in which context this can be called.
Your suggestion (below) to have the interrupts masked from this routine still makes sense.

Regarding page fault, won't pager execute in abort context and restore the cpu whatever its SVC/SYS state?

Reply:
It seems likely that this will be handled correctly by the pager. 
Regarding foreign interrupts, it looks like those should be OK but you'll depend on a very specific behaviour if foreign interrupts are received as FIQ. So masking all interrupts makes sense.

Reply:
Ok, covering both native/foreign int mapping is worth the game here.

Reply:
done